File: api/routes/simulation.py
"""
Simulation Routes - API endpoints for simulation control.

This module provides FastAPI routes for starting, stopping, and monitoring
MQTT simulations.
"""

from fastapi import APIRouter, HTTPException, Query

from ..models import (
    SimulationConfig,
    SimulationStatus,
    SimulationData,
    APIResponse,
)
from ..services.mqtt_service import mqtt_service

router = APIRouter(prefix="/simulation", tags=["simulation"])


@router.post("/start", response_model=APIResponse)
async def start_simulation(config: SimulationConfig):
    """Start a new MQTT simulation with the given configuration."""
    try:
        success = mqtt_service.start_simulation(config)

        if success:
            return APIResponse(
                success=True,
                message=(
                    f"Simulation started successfully with profile: "
                    f"{config.profile.name}"
                ),
                data={
                    "profile": config.profile.name,
                    "topic": config.profile.topic or "auto-generated",
                    "interval": config.interval,
                },
            )
        else:
            raise HTTPException(
                status_code=400,
                detail=(
                    "Failed to start simulation. Check MQTT connection "
                    "and configuration."
                ),
            )

    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Error starting simulation: {str(e)}"
        )


@router.post("/stop", response_model=APIResponse)
async def stop_simulation():
    """Stop the currently running simulation."""
    try:
        success = mqtt_service.stop_simulation()

        if success:
            return APIResponse(
                success=True, message="Simulation stopped successfully"
            )
        else:
            return APIResponse(
                success=False, message="No simulation is currently running"
            )

    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Error stopping simulation: {str(e)}"
        )


@router.get("/status", response_model=SimulationStatus)
async def get_simulation_status():
    """Get the current simulation status."""
    try:
        return mqtt_service.get_simulation_status()
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error getting simulation status: {str(e)}",
        )


@router.get("/data", response_model=SimulationData)
async def get_simulation_data(
    limit: int = Query(
        default=50, ge=1, le=100, description="Number of messages to return"
    )
):
    """Get recent simulation data."""
    try:
        data = mqtt_service.get_simulation_data(limit=limit)
        return SimulationData(**data)
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Error getting simulation data: {str(e)}"
        )

File: frontend/src/lib/stores/simulation.js
/**
 * Simulation Store
 * Manages the state of MQTT simulation
 */

import { writable } from 'svelte/store';
import { apiClient } from '../api.js';

// Create stores
export const simulationStatus = writable({
    isRunning: false,
    isConnected: false,
    profileName: null,
    topic: null,
    interval: null,
    messagesSent: 0,
    startTime: null,
    lastMessage: null
});

export const simulationData = writable({
    messages: [],
    totalCount: 0
});

export const profiles = writable([]);
export const selectedProfile = writable(null);
export const profilePreview = writable(null);
export const profileParams = writable({});

export const mqttConfig = writable({
    host: 'test.mosquitto.org',
    port: 1883,
    username: '',
    password: '',
    keepalive: 60,
    cleanSession: true,
    qos: 1,
    retained: false,
    lastWillTopic: '',
    lastWillMessage: '',
    lastWillQos: 0
});

export const simulationConfig = writable({
    interval: 5,
    duration: null
});

export const connectionStatus = writable({
    isConnected: false,
    isTesting: false,
    error: null
});

// Actions
export const simulationActions = {
    async loadProfiles() {
        try {
            const response = await apiClient.getProfiles();
            profiles.set(response.profiles);
        } catch (error) {
            console.error('Failed to load profiles:', error);
            throw error;
        }
    },

    async selectProfile(profileName) {
        try {
            const response = await apiClient.getProfileInfo(profileName);
            selectedProfile.set(response);
            profileParams.set({ ...response.parameters });
            
            // Generate preview with default parameters
            const previewResponse = await apiClient.generateProfilePreview(profileName, {});
            profilePreview.set(previewResponse.data.preview);
        } catch (error) {
            console.error('Failed to select profile:', error);
            throw error;
        }
    },

    async generatePreview(profileName, parameters) {
        try {
            const response = await apiClient.generateProfilePreview(profileName, parameters);
            profilePreview.set(response.data.preview);
        } catch (error) {
            console.error('Failed to generate preview:', error);
            throw error;
        }
    },

    async testConnection() {
        connectionStatus.update(status => ({ ...status, isTesting: true, error: null }));
        try {
            const config = get(mqttConfig);
            // Conversione camelCase -> snake_case per compatibilità backend
            const apiConfig = {
                host: config.host,
                port: config.port,
                username: config.username,
                password: config.password,
                keepalive: config.keepalive,
                clean_session: config.cleanSession,
                qos: config.qos,
                retained: config.retained,
                last_will_topic: config.lastWillTopic || null,
                last_will_message: config.lastWillMessage || null,
                last_will_qos: config.lastWillQos || 0
            };
            const response = await apiClient.testMqttConnection(apiConfig);
            console.log("MQTT test response:", response);
            connectionStatus.update(status => ({
                ...status,
                isTesting: false,
                isConnected: response.success,
                error: response.success ? null : response.message
            }));
            return response.success;
        } catch (error) {
            connectionStatus.update(status => ({
                ...status,
                isTesting: false,
                isConnected: false,
                error: error.message
            }));
            throw error;
        }
    },

    async startSimulation() {
        try {
            const mqtt = get(mqttConfig);
            const sim = get(simulationConfig);
            const profile = get(selectedProfile);
            
            if (!profile) {
                throw new Error('No profile selected');
            }

            const config = {
                mqtt,
                profile: {
                    name: profile.name,
                    parameters: profile.parameters || {},
                    topic: null // Use default topic
                },
                interval: sim.interval,
                duration: sim.duration
            };

            const response = await apiClient.startSimulation(config);
            
            if (response.success) {
                // Update status
                await this.updateStatus();
            }
            
            return response;
        } catch (error) {
            console.error('Failed to start simulation:', error);
            throw error;
        }
    },

    async stopSimulation() {
        try {
            const response = await apiClient.stopSimulation();
            
            if (response.success) {
                // Update status
                await this.updateStatus();
            }
            
            return response;
        } catch (error) {
            console.error('Failed to stop simulation:', error);
            throw error;
        }
    },

    async updateStatus() {
        try {
            const status = await apiClient.getSimulationStatus();
            simulationStatus.set({
                isRunning: status.is_running,
                isConnected: status.is_connected,
                profileName: status.profile_name,
                topic: status.topic,
                interval: status.interval,
                messagesSent: status.messages_sent,
                startTime: status.start_time,
                lastMessage: status.last_message
            });
        } catch (error) {
            console.error('Failed to update status:', error);
        }
    },

    async loadSimulationData(limit = 50) {
        try {
            const data = await apiClient.getSimulationData(limit);
            simulationData.set({
                messages: data.messages,
                totalCount: data.total_count
            });
        } catch (error) {
            console.error('Failed to load simulation data:', error);
        }
    }
};

// Helper function to get store value
function get(store) {
    let value;
    store.subscribe(v => value = v)();
    return value;
}

// Auto-update status when simulation is running
let statusInterval;

simulationStatus.subscribe(status => {
    if (status.isRunning) {
        if (!statusInterval) {
            statusInterval = setInterval(() => {
                simulationActions.updateStatus();
                simulationActions.loadSimulationData();
            }, 1000);
        }
    } else {
        if (statusInterval) {
            clearInterval(statusInterval);
            statusInterval = null;
        }
    }
}); 